include/nil/crypto3/zk/snark/systems/plonk/placeholder/permutation_argument.hpp

```diff
@@ -69,17 +69,14 @@ namespace nil {
                     };
 
                     static inline math::polynomial_dfs<typename FieldType::value_type> polynomial_product(
-                        std::vector<math::polynomial_dfs<typename FieldType::value_type>> multipliers)
+                        std::vector<math::polynomial_dfs<typename FieldType::value_type>>& multipliers)
                     {
-                        while (multipliers.size() != 1) {
-                            for (int i = 0; i < multipliers.size() / 2; ++i) {
-                                multipliers[i] = multipliers[2 * i] * multipliers[2 * i + 1];
+                        std::size_t stride = 1;
+                        while (stride < multipliers.size() ) {
+                            for(std::size_t i = 0; i + stride < multipliers.size(); i += stride*2) {
+                                multipliers[i] *= multipliers[ i + stride ];

```

Comment:
Will this work, if the size of multipliers is not a power of 2?

Reply:
Sure it will. non-multiples will accumulate in last accounted element, e.g. for 7:
```
1 1 1 1 1 1 1
2   2   2   1
4       3
7
```